Use foreground color in macOS tab titles

BlakeWilliams opened the pull request:
I'm not a Zig or a Swift/macOS developer but I've been really liking GhostTTY and wanted to open this PR to either contribute back, or at least share the progress/learnings when tackling https://github.com/ghostty-org/ghostty/issues/2764 (which I think this should close, if this is a valid approach). I have a vimscript that syncs the terminal colors with my vim theme, so I was pretty motivated to implement this feature.

This change was relatively straightforward (but required a lot of learning). In general it:

- Tracks `foregroundColor` in the SurfaceView. It's pre-populated with the config defined foreground color. It is also updated via the color change events (which I added foreground support to)
- Adds a new `titleForegroundColor` property to use in `TerminalWindow` titles so the colors can be updated dynamically.
- Updates each window's tab foreground color when `syncAppearance` is called. (I'm honestly not sure if this is the best approach, but it's the solution I found. happy to update if there's a better approach)
- Calls `syncAppearance` when a window becomes the key window to sync the colors across each tab.
- Updates some fonts to use 11pt sizes instead of `systemFontSize` to match the existing behavior/size of tabs (happy to provide screenshots to elaborate)

Here's a quick screencap of the new behavior:

https://github.com/user-attachments/assets/ba97432e-193a-4afd-95cd-faf23694530a


<details>
  <summary>I also quickly validated that this works with `macos-titlebar-style=transparent` in addition to `tabs`</summary>

<img width="912" alt="Screenshot 2024-12-28 at 10 01 24 PM" src="https://github.com/user-attachments/assets/42447da7-3661-418c-8cd5-7e41430c9591" />
</details>

<details>
  <summary>and just for completions sake, bg transparency:</summary>
  <img width="819" alt="Screenshot 2024-12-28 at 10 04 20 PM" src="https://github.com/user-attachments/assets/293e1041-79ff-4a3c-9d52-47f24e0880d4" />
</details>

BlakeWilliams committed:
Use foreground color in macOS tab titles

Currently tab titles will use the default macOS tab foreground color,
but the color schemes background color. This can result in tabs with
unreadable text.

This change updates the macOS UI tabs to use the foreground color
defined in the config, or the overridden value when set dynamically.

This change:

- Tracks `foregroundColor` in the SurfaceView which is pre-populated
  with the config foreground color along with support for color change
  update events.
- Adds a new `titleForegroundColor` property to use in `TerminalWindow`
  titles.
- Updates each tab window foreground color when `syncAppearance` is
  called.
- Calls `syncAppearance` when a window becomes the key window.

mitchellh commented on macos/Sources/Features/Terminal/TerminalController.swift:
```diff
@@ -268,6 +273,11 @@ class TerminalController: BaseTerminalController {
+        guard let windows = self.window?.tabbedWindows as? [TerminalWindow] else { return }
+        for (w) in windows {
+            w.titleForegroundColor = window.titleForegroundColor
+        }
```
It feels like this should be moved to a private func and this should be called here and in `windowDidBecomeKey` instead of a _full `syncAppearance`_ which is rather expensive. Thoughts?

BlakeWilliams replied:
That sounds good and reasonable to me.  I'll make that change later today and do some more manual testing to make sure all the existing cases are still handled.

BlakeWilliams replied:
Updated in https://github.com/ghostty-org/ghostty/pull/3843/commits/3c17b9176e6c6cbfca6b83ce5943c5b185480ebb

mitchellh reviewed (changes requested):
Honestly, very good job. I have one question/request.

BlakeWilliams committed:
Extract window title color updates into private function

This extracts the window title update logic into a new function so we
can avoid calling syncAppearance on `windowDidBecomeKey` which is much
heavier than updating just the titles.

BlakeWilliams committed:
Move guard as well

BlakeWilliams commented:
I noticed this while testing another color scheme, but some color schemes with light backgrounds can still be difficult to read. I think this is due to lighter tab backgrounds being shaded when dimmed.

Happy to address it in this PR if necessary, but also happy to try finding some time to look into it as a follow-up if it's not a blocker. 

<img width="750" alt="Screenshot 2025-01-08 at 10 32 38 PM" src="https://github.com/user-attachments/assets/12218ae2-be62-41a3-9590-a8b42545c61a" />
